### PR TITLE
early stopping for CV (python)

### PR DESCRIPTION
**Pull request by Far0n**
- implements early_stopping_rounds parameter for cv
issue https://github.com/dmlc/xgboost/issues/529#issuecomment-154198031
- unittest

### CODE REVIEW

File: python-package/xgboost/training.py [hunk 1]
```diff
@@ -332,6 +338,27 @@ def cv(params, dtrain, num_boost_round=10, nfold=3, metrics=(),
     -------
     evaluation history : list(string)
     """
+    if early_stopping_rounds is not None and len(metrics) > 1:
+        raise ValueError('Check your params.'\
+                                     'Early stopping works with single eval metric only.')
+
+    if early_stopping_rounds is not None:
+        sys.stderr.write("Will train until cv error hasn't decreased in {} rounds.\n".format(\
```
Comment (terrytangyuan) [inline python-package/xgboost/training.py]: You might want to combine the above if statements for `early_stopping_rounds` and then have an inner if statement for `len(metrics) > 1`. Also wrap the `sys.stderr.write` around with `if verbose_eval` like I did in [this commit](15a0d27).

File: python-package/xgboost/training.py [hunk 2]
```diff
@@ -342,6 +369,18 @@ def cv(params, dtrain, num_boost_round=10, nfold=3, metrics=(),
...
 
+        if early_stopping_rounds is not None:
+            score = res[0]
+            if (maximize_score and score > best_score) or \
+                    (not maximize_score and score < best_score):
+                best_score = score
+                best_score_i = i
+            elif i - best_score_i >= early_stopping_rounds:
+                sys.stderr.write("Stopping. Best iteration: {}\n".format(best_score_i))
+                best_score = best_score
```
Comment (terrytangyuan) [inline python-package/xgboost/training.py]: What does this line do? Also maybe a better naming for `best_score_i`? This `stderr.write` can be left as it is (not like the previous one).

File: python-package/xgboost/training.py (hunk 1 above)
Comment (Far0n) [inline python-package/xgboost/training.py]: cv does not have a `verbose_eval` parameter. Shall I use `show_progress` as precondition instead?

**Comment by Far0n** [comment]
"*Also maybe a better naming for best_score_i?*" 
- I used the same name as in `train()` for consistency

"*Also wrap the sys.stderr.write around with if verbose_eval like I did in this commit.*"
- cv does not have a `verbose_eval` parameter. Shall I use `show_progress` as precondition instead? Introducing `verbose_eval` could be misleading here, but using `show_progress` leads also to another behavior compared to `train`, because `verbose_eval` defaults to `True` whereas `show_progress` defaults to `False`. Hence, I don't know what's best here.

**Comment by terrytangyuan** [comment]
Ok, let's leave it as it is. Merged. Thanks!

**Comment by terrytangyuan** [comment]
@Far0n I wasn't paying attention but it seems that the unit test disappeared?

**Comment by Far0n** [comment]
@terrytangyuan Sry my fault. I pushed from the wrong origin .. I just created a PR with the "lost" unittest.

**Pull request merged**